build: fix ci cacheing to use recently built caches

kormide opened the pull request:
@alexeagle

Github actions caches don't update already existing caches, so the cache that gets loaded could be old. This sets a new cache using a cache key derived from a hash of files. The use of `restore-keys: bazel-cache-` will find the most recent cache stored scoped to this branch or a parent branch.

kormide commented on .github/workflows/release.yml:
```diff
@@ -14,6 +14,12 @@ jobs:
     steps:
       - name: Checkout
         uses: actions/checkout@v2
+      - name: Mount bazel repo cache
+        uses: actions/cache@v2
+        with:
+          path: "~/.cache/bazel-repo"
+          key: bazel-cache-${{ hashFiles('**/BUILD.bazel', '**/*.bzl', 'WORKSPACE') }}
+          restore-keys: bazel-cache-
```
My assumption for a release build is that we can re-use the repository cache but not the action cache in case there's stamping. Does that seem right?

alexeagle commented on .github/workflows/ci.yaml:
```diff
@@ -20,16 +20,14 @@ jobs:
+          path: |
+            "~/.cache/bazel"
+            "~/.cache/bazel-repo"
+          key: bazel-cache-${{ hashFiles('**/BUILD.bazel', '**/*.bzl', 'WORKSPACE') }}
```
why do we need a cache key here? can't we always restore the cache regardless of the hashes of certain files, and rely on bazel to ignore outdated entries in it?

kormide replied:
As we discussed offline, once a cache is stored it can never be updated, so we need to write under new cache keys whenever things change. See comment I added for clarification.

f0rmiga commented on .github/workflows/ci.yaml:
```diff
@@ -20,16 +20,21 @@ jobs:
-      - name: Mount bazel action cache
+      # Cache build and external artifacts so that the next ci build is incremental.
+      # Because caches cannot be updated, create a new cache based on the contents of
+      # hashFiles (modify this to capture sources in your repo that would cause bazel
```
I think you could add another step to the root README.md, pointing the user here.

kormide replied:
Added

kormide commented on .github/workflows/ci.yaml:
```diff
@@ -20,16 +20,21 @@ jobs:
+      # Cache build and external artifacts so that the next ci build is incremental.
+      # Because caches cannot be updated, create a new cache based on the contents of
+      # hashFiles (modify this to capture sources in your repo that would cause bazel
+      # to rebuild artifacts, but it doesn't have to be perfect). On a cache miss, the
```
Sources that are inputs to bazel. To minimize the amount of build work that would mean every source. But if it's onerous to declare every source in the hashFiles then it's not the end of the world to leave some out. You want enough sources so that the bulk of the bazel work will be cached. I'll try to reword.

kormide commented:
Updated. @alexeagle I wrote a more verbose (but hopefully clearer) explanation. @f0rmiga I removed the cache location flags outside of the bazelrc file as it doesn't support environment variables.
